### PR TITLE
Document how to create new cloud provider API and validate requests

### PR DESCRIPTION
**Pull request by deitch**
**What this PR does / why we need it**:
Document the steps to add a new cloud provider to the Gardener API and validate the requests

**Special notes for your reviewer**:
As discussed with @rfranzke on Slack. Continuation of #647 

**Release note**:

```improvement developer
Document adding new cloud provider to gardener component
```

### CODE REVIEW

**Comment by deitch** [comment]
It feels like there is a lot of repetition here, @rfranzke . Does some of the code need to be created twice? Once at high-level and once in `v1beta1`?

**Comment by rfranzke** [comment]
What do you mean? I didn't get it

**Comment by deitch** [comment]
> I didn't get it

I based the doc on #360 and #366. If I look at #360, I see a lot of what goes into https://github.com/gardener/gardener/blob/master/pkg/apis/garden/types.go is similar to or duplicative of what goes into https://github.com/gardener/gardener/blob/master/pkg/apis/garden/v1beta1/types.go . Same for `helpers.go`.

**Comment by rfranzke** [comment]
Can you give an example?

**Comment by deitch** [comment]
Sure. Using #360, look at [here](https://github.com/gardener/gardener/pull/360/files#diff-9c42317eee788d58380acadcfab41545R47) and [here](https://github.com/gardener/gardener/pull/360/files#diff-a3d3f75ad34298fb9696284b0ef772c2R55). There were a few more.

**Comment by rfranzke** [comment]
Ah OK, yes, the types are "duplicate" (not in general, but in our case). The internal API version is the same like the external API version (v1beta1).
See also https://github.com/kubernetes/kubernetes/blob/master/pkg/apis/apps/types.go and https://github.com/kubernetes/api/blob/master/apps/v1/types.go, that's the same.

The API server acts on the internal version while the controller-manager acts on the v1beta1 version.

**Comment by deitch** [comment]
Yeah, makes sense.

I will await your review (i.e. finding my errors) on the rest of the doc.

File: docs/development/new-cloud-provider-api-extension.md [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+1. Add the necessary structures, logic and tests to the code
+2. Update example templates
+3. Generate auto-generated code
+4. Generate examples
+
+The detailed steps are as follows:
+
+1. Add the cloud provider to the template files in [hack/templates/resources](../../hack/templates/resources/) :
+  * `CloudProfile` [30-cloudprofile.yaml.tpl](../../hack/templates/resources/30-cloudprofile.yaml.tpl) This is the definition of a profile for this provider.
+  * `Secret` [40-secret-seed.yaml.tpl](../../hack/templates/resources/40-secret-seed.yaml.tpl) This contains the secrets necessary to invoke the cloud provider's API to deploy the Seed cluster, and the kubeconfig to access the Seed cluster.
```
Comment (rfranzke) [inline docs/development/new-cloud-provider-api-extension.md]: Regarding the review, I thought you are still working on it. About this part we already talked in Slack: The seed cluster is not deployed by Gardener, it's just connected to it by providing the proper kubeconfig.

Comment (deitch) [reply docs/development/new-cloud-provider-api-extension.md]: Oh yes, I missed updating this. Coming now.

Comment (deitch) [reply docs/development/new-cloud-provider-api-extension.md]: Updated.

Comment (rfranzke) [reply docs/development/new-cloud-provider-api-extension.md]: Thx! Will check soon 👍

File: docs/development/new-cloud-provider-api-extension.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+Cloud provider support is headed through the following major steps:
+
+1. Add the necessary structures, logic and tests to the code
+2. Update example templates
+3. Generate auto-generated code
+4. Generate examples
+
+The detailed steps are as follows:
+
+1. Add the cloud provider to the template files in [hack/templates/resources](../../hack/templates/resources/) :
```
Comment (rfranzke) [inline docs/development/new-cloud-provider-api-extension.md]: s/the cloud provider/the cloud provider specific configuration

File: docs/development/new-cloud-provider-api-extension.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+  * `CloudProfile` [30-cloudprofile.yaml.tpl](../../hack/templates/resources/30-cloudprofile.yaml.tpl) This is the definition of a profile for this provider.
+  * `Secret` [40-secret-seed.yaml.tpl](../../hack/templates/resources/40-secret-seed.yaml.tpl) This contains the secrets necessary to invoke the cloud provider's API to manage Shoot ectd backups, and the kubeconfig to access the Seed cluster.
```
Comment (rfranzke) [inline docs/development/new-cloud-provider-api-extension.md]: s/to manage .../to create blob storage buckets for managing the etcd backups of shoot clusters

File: docs/development/new-cloud-provider-api-extension.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+2. Update example templates
+3. Generate auto-generated code
+4. Generate examples
+
+The detailed steps are as follows:
+
+1. Add the cloud provider to the template files in [hack/templates/resources](../../hack/templates/resources/) :
+  * `CloudProfile` [30-cloudprofile.yaml.tpl](../../hack/templates/resources/30-cloudprofile.yaml.tpl) This is the definition of a profile for this provider.
+  * `Secret` [40-secret-seed.yaml.tpl](../../hack/templates/resources/40-secret-seed.yaml.tpl) This contains the secrets necessary to invoke the cloud provider's API to manage Shoot ectd backups, and the kubeconfig to access the Seed cluster. 
+  * `Seed`: [50-seed.yaml.tpl](../../hack/templates/resources/50-seed.yaml.tpl) This contains the configuration for the Seed cluster..
```
Comment (rfranzke) [inline docs/development/new-cloud-provider-api-extension.md]: s/cluster../cluster.

File: docs/development/new-cloud-provider-api-extension.md [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+3. Generate auto-generated code
+4. Generate examples
+
+The detailed steps are as follows:
+
+1. Add the cloud provider to the template files in [hack/templates/resources](../../hack/templates/resources/) :
+  * `CloudProfile` [30-cloudprofile.yaml.tpl](../../hack/templates/resources/30-cloudprofile.yaml.tpl) This is the definition of a profile for this provider.
+  * `Secret` [40-secret-seed.yaml.tpl](../../hack/templates/resources/40-secret-seed.yaml.tpl) This contains the secrets necessary to invoke the cloud provider's API to manage Shoot ectd backups, and the kubeconfig to access the Seed cluster. 
+  * `Seed`: [50-seed.yaml.tpl](../../hack/templates/resources/50-seed.yaml.tpl) This contains the configuration for the Seed cluster.. 
+  * `Secret`: [70-secret-cloudprovider.yaml.tpl](../../hack/templates/70-secret-cloudprovider.yaml.tpl) This contains the secrets necessary for to invoke the cloud provider's API to deploy the Shoot cluster(s).
```
Comment (rfranzke) [inline docs/development/new-cloud-provider-api-extension.md]: s/secrets/credentials

File: docs/development/new-cloud-provider-api-extension.md [hunk 6]
```diff
@@ -0,0 +1,79 @@
...
+  * `SecretBinding`: [80-secretbinding-cloudprovider.yaml.tpl](../../hack/templates/resources/80-secretbinding-cloudprovider.yaml.tpl) This binds the secret defined in the previous step with quotas from the current or another namespace.
```
Comment (rfranzke) [inline docs/development/new-cloud-provider-api-extension.md]: I would leave the quota part out to reduce surface for confusion - we will provide separate documentation on how exactly the `SecretBinding` resource is used. Let's focus here on the cloud provider extension part.

File: docs/development/new-cloud-provider-api-extension.md [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+
+Validations check that parameters submitted to Gardener for the given cloud profile are valid. It has two major steps:
+
+1. Write validators
+2. Write tests for the validators
+
+The details are as follows:
+
+1. Update [plugin/pkg/shoot/quotavalidator/admission.go](../../plugin/pkg/shoot/quotavalidator/admission.go) at the following key points:
+    * Add logic in `getShootWorkerResources()` to populate the return `[]quoteworker` slice with the workers for the cloud provider.
```
Comment (rfranzke) [inline docs/development/new-cloud-provider-api-extension.md]: s/quoteworker/quotaworker

File: docs/development/new-cloud-provider-api-extension.md [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+    * Add logic in `getMachineTypes()` to return the list of machine types available for the cloud provider.
+    * Add logic in `getVolumeTypes()` to return the list of volume types available for the cloud provider.
+    * Add logic in `quotaVerificationNeeded()` to return if quote verification is required. Note: What is "quote verification"?
```
Comment (rfranzke) [inline docs/development/new-cloud-provider-api-extension.md]: s/quote/quota

It is possible to define Quotas for Shoot clusters, e.g., project "A" may use shoots with max. 8 CPUs.
The "quota verification" function checks whether a shoot that was put under quota has changes in relevant sections of the shoot specification that might be related to CPU (in this example), e.g. the worker pools. The admission controller checks whether this change is allowed or not (constrained by the defined Quota).

File: docs/development/new-cloud-provider-api-extension.md [hunk 9]
```diff
@@ -0,0 +1,79 @@
...
+2. Update [plugin/pkg/shoot/quotavalidator/admission_test.go](../../plugin/pkg/shoot/quotavalidator/admission_test.go) to test logic from the previous step.
+3. Update [plugin/pkg/shoot/validator/admission.go](../../plugin/pkg/shoot/validator/admission.go) at the following key points:
+    * Add logic to `Admit()` where other cloud providers exist to populate the keys of `&garden.Cloud{}` with the `MachineImage` for the cloud provider. See existing examples.
+    * Add logic to `switch cloudProviderInShoot` to add a case for the cloud provider, validating that it can get the machine image. The logic here should be minimal, instead calling out to a function that does more detailed validation, e.g. `validateAWS()` or `validateAzure()`, i.e. `validate<Provider>()`.
+    * Create the function `validate<Provider>()` that validates all the elements of the request, including sub-functions, if necessary. See existing examples.
+4. Update [plugin/pkg/shoot/validator/admission_test.go](../../plugin/pkg/shoot/validator/admission_test.go) to test logic from the previous step.
+
```
Comment (rfranzke) [inline docs/development/new-cloud-provider-api-extension.md:77]: Remove unneeded empty lines at the end of the document

**Review by rfranzke — CHANGES_REQUESTED** [review]
Very detailed, thank you for your efforts. Once we move out the cloud providers we will need to adapt the document, but the degree of detail shows that the move-out is very much needed. We must depend on clean APIs instead of munching everything together as of today. Thanks again!

File: docs/development/new-cloud-provider-api-extension.md (hunk 6 above)
Comment (deitch) [reply docs/development/new-cloud-provider-api-extension.md]: Hmm... ok. But we do need to update that `tpl` file, so some description would be useful. I guess as long as it binds to the right `secretRef` we are good.

File: docs/development/new-cloud-provider-api-extension.md (hunk 7 above)
Comment (deitch) [reply docs/development/new-cloud-provider-api-extension.md]: Do you want to quote that? :-)

Comment (rfranzke) [reply docs/development/new-cloud-provider-api-extension.md]: [?](https://github.com/gardener/gardener/blob/master/plugin/pkg/shoot/quotavalidator/admission.go#L430)

**Comment by deitch** [comment]
> the degree of detail shows that the move-out is very much needed

Yes, indeed. It took quite some time to back out all of the areas. Even having written this, I am a little nervous about doing the implementation, but we shall see.

> We must depend on clean APIs instead of munching everything together as of today.

You have my vote! :-)

Addressed all of your comments and rebased, of course. Back to you.

File: docs/development/new-cloud-provider-api-extension.md (hunk 7 above)
Comment (deitch) [reply docs/development/new-cloud-provider-api-extension.md]: I was making a pun on my type of "quota" to "quote". The pun failed...

Comment (rfranzke) [reply docs/development/new-cloud-provider-api-extension.md]: Haha 👍

Comment (deitch) [reply docs/development/new-cloud-provider-api-extension.md]: It was a pretty bad joke anyways...

**Comment by rfranzke** [comment]
Thank's, looks good so far. PR is still labelled as "WIP" - should we wait with merge?

**Comment by deitch** [comment]
Not at all. I put `WIP` more as an indication of, “I’m not sure what I’m doing here, and it probably will take a few reviews by @rfranzke and others before it makes any sense!” :-)

I’ll remove that now.

**Comment by deitch** [comment]
Is there anything else to add? Did I miss any?

**Comment by rfranzke** [comment]
Don't think so. Let's merge it. :) Thank you again!

**Review by rfranzke — APPROVED** [review]
/lgtm

**Pull request merged**